### PR TITLE
[Merged by Bors] - Add apply_or_insert functions to reflected component and resources

### PR DESCRIPTION
**Pull request by Shatur**
# Objective

`ReflectResource` and `ReflectComponent` will panic on `apply` method if there is no such component. It's not very ergonomic. And not very good for performance since I need to check if such component exists first.

## Solution

* Add `ReflectComponent::apply_or_insert` and `ReflectResource::apply_or_insert` functions.
* Rename `ReflectComponent::add` into `ReflectComponent::insert` for consistency.

---

### Added

* `ReflectResource::apply_or_insert` and `ReflectComponent::apply_on_insert`.

### Changed

* Rename `ReflectComponent::add` into `ReflectComponent::insert` for consistency.
* Use `ReflectComponent::apply_on_insert` in `DynamicScene` instead of manual checking.

## Migration Guide

* Rename `ReflectComponent::add` into `ReflectComponent::insert`.

### CODE REVIEW

File: crates/bevy_ecs/src/reflect.rs [hunk 1]
```diff
@@ -45,6 +46,20 @@ impl ReflectComponent {
         (self.apply_component)(world, entity, component);
     }
 
+    /// Uses reflection to set the value of this [`Component`] type in the entity to the given value or insert a new one.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/reflect.rs:49]:
```suggestion
    /// Uses reflection to set the value of this [`Component`] type in the entity to the given value or insert a new one if it does not exist.
```

File: crates/bevy_ecs/src/reflect.rs [hunk 2]
```diff
@@ -183,6 +208,11 @@ impl ReflectResource {
         (self.apply_resource)(world, resource);
     }
 
+    /// Uses reflection to set the value of this [`Resource`] type in the world to the given value or insert a new one.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/reflect.rs:211]:
```suggestion
    /// Uses reflection to set the value of this [`Resource`] type in the world to the given value or insert a new one if it does not exist.
```

**Review by alice-i-cecile — APPROVED** [review]
LGTM once the doc comment changes are in.

**Comment by Shatur** [comment]
While working on it I decided to open an issue with naming suggestions: #5202

**Comment by Shatur** [comment]
@alice-i-cecile rebased after https://github.com/bevyengine/bevy/pull/5219.

**Review by PROMETHIA-27 — APPROVED** [review]

**Pull request closed without merging**